src/panels/ActionsPanel/ActionsPanel.tsx

```diff
@@ -99,52 +167,117 @@ export default function ActionsPanel(): JSX.Element {
           Run action on {appName}: {generateSelectedUnitList()}
         </div>
         <div className="actions-panel__action-list">
-          {generateActionlist(
-            actionData,
-            fetchingActionData,
-            selectedAction,
-            setSelectedAction
-          )}
+          <LoadingHandler
+            data={data}
+            loading={fetchingActionData}
+            noDataMessage="This charm has not provided any actions."
+          >
+            {Object.keys(actionData).map((actionName) => (
+              <RadioInputBox
+                name={actionName}
+                description={actionData[actionName].description}
+                onSelect={selectHandler}
+                selectedInput={selectedAction}
+                key={actionName}
+              >
+                <ActionOptions
+                  name={actionName}
+                  data={actionData}
+                  onValuesChange={changeHandler}
+                />
+              </RadioInputBox>
+            ))}
+          </LoadingHandler>
+        </div>
+        <div className="actions-panel__drawer">
+          <button
+            className={classNames(
+              "p-button--positive actions-panel__run-action"
+            )}
+            disabled={disableSubmit}
+            onClick={executeAction}
+          >
+            Run action
+          </button>
         </div>
       </div>
     </Aside>
   );
 }
 
-function generateActionlist(
-  actionData: ActionData | undefined,
-  fetchingActionData: boolean,
-  selectedAction: string | undefined,
-  setSelectedAction: SetSelectedAction
+function onValuesChange(
+  actionName: string,
+  values: ActionOptionValue,
+  optionValues: MutableRefObject<ActionOptionValues>
 ) {
-  if (!actionData && fetchingActionData) return null;
-  if (!actionData && !fetchingActionData)
-    return <div>This charm has not provided any actions</div>;
-  if (!actionData) return null;
-
-  return Object.keys(actionData).map((actionName) => {
-    const action = actionData[actionName];
-    const options: ActionOptions = [];
-
-    Object.keys(action.params.properties).forEach((name) => {
-      const property = action.params.properties[name];
-      options.push({
-        name: name,
-        description: property.description,
-        type: property.type,
-        required: action.params.required.includes(name),
-      });
-    });
-
-    return (
-      <RadioInputBox
-        name={actionName}
-        description={action.description}
-        options={options}
-        onSelect={setSelectedAction}
-        selectedAction={selectedAction}
-        key={actionName}
-      />
-    );
+  const updatedValues: ActionOptionValue = {};
+  Object.keys(values).forEach((key) => {
+    updatedValues[key.replace(`${actionName}-`, "")] = values[key];
   });
+
+  optionValues.current = {
+    ...optionValues.current,
```

Comment:
Why do you need the old action's/actions' values here? Don't you only care about the new ones? I might not be following something...

Reply:
Nope, I technically don't need the old actions information but it's nice to have it. 

Technically a person can commit multiple actions at once (we don't currently support that). So in the future I won't need to change this code when we want to enable support for that feature.